src/Libraries/Nop.Core/Caching/MemoryCacheManager.cs

```diff
@@ -42,47 +40,36 @@ public MemoryCacheManager(AppSettings appSettings, IMemoryCache memoryCache) : b
         /// </summary>
         /// <param name="key">Cache key</param>
         /// <returns>Cache entry options</returns>
-        private MemoryCacheEntryOptions PrepareEntryOptions(CacheKey key)
+        private static MemoryCacheEntryOptions PrepareEntryOptions(CacheKey key)
         {
             //set expiration time for the passed cache key
             var options = new MemoryCacheEntryOptions
             {
                 AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(key.CacheTime)
             };
 
-            //add tokens to clear cache entries
+            //add token to clear cache entries
             options.AddExpirationToken(new CancellationChangeToken(_clearToken.Token));
-            foreach (var keyPrefix in key.Prefixes.ToList())
-            {
-                var tokenSource = _prefixes.GetOrAdd(keyPrefix, new CancellationTokenSource());
-                options.AddExpirationToken(new CancellationChangeToken(tokenSource.Token));
-            }
+            options.RegisterPostEvictionCallback(OnEviction);
+            _keys.Add(key.Key, default);
 
             return options;
         }
 
-        /// <summary>
-        /// Remove the value with the specified key from the cache
-        /// </summary>
-        /// <param name="cacheKey">Cache key</param>
-        /// <param name="cacheKeyParameters">Parameters to create cache key</param>
-        private void Remove(CacheKey cacheKey, params object[] cacheKeyParameters)
-        {
-            cacheKey = PrepareKey(cacheKey, cacheKeyParameters);
-            _memoryCache.Remove(cacheKey.Key);
-        }
-
-        /// <summary>
-        /// Add the specified key and object to the cache
-        /// </summary>
-        /// <param name="key">Key of cached item</param>
-        /// <param name="data">Value for caching</param>
-        private void Set(CacheKey key, object data)
+        private static void OnEviction(object key, object value, EvictionReason reason, object state)
         {
-            if ((key?.CacheTime ?? 0) <= 0 || data == null)
-                return;
-
-            _memoryCache.Set(key.Key, data, PrepareEntryOptions(key));
+            switch (reason)
+            {
+                // we clean up after ourselves elsewhere
+                case EvictionReason.Removed:

```

Comment:
Note that a key might be removed by the Redis-synced cache and thus wouldn't be removed from _keys here. That would by no means be a big problem (worst case, RemoveByPrefix would try to remove a few nonexistent keys), but I'll do something about it anyway for the sake of cleanliness.

Reply:
Resolved by commit [9f5b144](https://github.com/nopSolutions/nopCommerce/pull/6531/commits/9f5b1445f996e1d3daa8f130cbe16dd251e44f8a)